### PR TITLE
feat(s3stream): support limit write bandwidth for WALBlockDeviceChannel

### PR DESCRIPTION
**Pull request by lifepuzzlefun**
### why need this?
1. the block device provider have limit on iobandwidth and iops. either one exceed the limit the write block channel time goes very high (cause by the block device qos)
2. when single node write request number is high. the write may exceed the limit iobandwidth for a little time but can cause the pending request accumlate and finally the broker will have very long write latency.

### CODE REVIEW

**Comment by Chillax-0v0** [comment]
Hello, thank you for your contribution. However, could you please explain the reasons for doing so?

**Comment by lifepuzzlefun** [comment]
image
our disk the upper limit is 210MB/s and 2800 iops. but the incoming request size maybe larger then one block so the iops not exceed but the write bandwidth exceeded.

File: s3stream/src/main/java/com/automq/stream/s3/wal/util/WALBlockDeviceChannel.java [hunk 1]
```diff
@@ -285,6 +299,15 @@ private void unalignedWrite(ByteBuf src, long position) throws IOException {
     private int write(ByteBuffer src, long position) throws IOException {
         assert WALUtil.isAligned(src.remaining());
 
+        int permits = src.remaining();
+        if (writeBandwidthLimit != DISABLE_WRITE_BAND_WIDTH_LIMIT && permits > 0) {
```
Comment (superhx) [inline s3stream/src/main/java/com/automq/stream/s3/wal/util/WALBlockDeviceChannel.java:303]: Default use Long.MAX_VALUE? Reduce branching logic.

**Comment by lifepuzzlefun** [comment]
code update, notice that bucket4j limit max 1token/ns which is 953.67MB/s, and when acquire 0 exception will be throwed.

**Comment by superhx** [comment]
> code update, notice that bucket4j limit max 1token/ns which is 953.67MB/s, and when acquire 0 exception will be throwed.

Should we throw exception when the bandwidth config exceeds the limitation or support higher limitation?

**Comment by lifepuzzlefun** [comment]
> > code update, notice that bucket4j limit max 1token/ns which is 953.67MB/s, and when acquire 0 exception will be throwed.
> 
> Should we throw exception when the bandwidth config exceeds the limitation or support higher limitation?

@superhx code changed to support 3814.7 MB/s. and when consume the bytes will >> 4. when user config larger than this exception will throw.

**Review by superhx — APPROVED** [review]

**Pull request merged**